Implementing the manager to invoke a REST service (Part 5)

elabcius5 opened the pull request:
For this PR, I implemented a manager to invoke a REST service:
- I used a several libraries (otto, okhttp, retrofit, json) for invoke a service through events
- I created a packages for a different classes for major order

![part 5](https://cloud.githubusercontent.com/assets/20821467/17997085/0f212598-6b32-11e6-95e3-9193c48dc91c.png)

elabcius5 committed:
Library Picasso

For this fourth delivery, I used the Picasso library and this library
loads images API Octokit

elabcius5 committed:
Merge branch 'master' into SubscribersSearch_FourthPR-

elabcius5 committed:
Changes for it shows a Dialog

elabcius5 committed:
Fix the merge errors

This commit fixes the many errors, when I merge the master into this
branch

elabcius5 committed:
Implement the manager to invoke a REST service

elabcius5 committed:
Correction a manager and calling to fragment

I changed a manager to consume a REST service, I deleted a class
application and I moved a logic to fragment class.
I changed a type to call a fragment, in MainActivity class, I add logic
for call transaction a fragment.

lucaslabs commented on app/src/main/java/com/globant/tutoria/subscriberssearch/manager/SubscribersManager.java:
```diff
@@ -0,0 +1,42 @@
+        sSubscribersClient.getSubscribersService().getSubscribers().enqueue(new Callback<List<SubscribersModel>>() {
+            @Override
+            public void onResponse(Call<List<SubscribersModel>> call, Response<List<SubscribersModel>> response) {
+                mBus.post(new SendSubscribersEvent(response.body()));
```
An event bus is not needed as discussed on checkpoint meeting. You can just simply have the callbacks (`onResponse` and `onFailure`) in the fragment where you use the `SubscribersManager`.

elabcius5 replied:
I edited the manager class and I fixed the service called

lucaslabs commented on app/src/main/java/com/globant/tutoria/subscriberssearch/fragments/SubscriberListFragment.java:
```diff
@@ -0,0 +1,83 @@
+        View rootView = inflater.inflate(R.layout.subscribers_list_fragment, container, false);
+        rootView.setTag(TAG);
+        recyclerViewSubscribers = (RecyclerView) rootView.findViewById(R.id.recycler_view_subscribers);
+        progressDialog.setMessage("Loading subscribers");
```
Magic String! Please create a resouce for this String.

elabcius5 replied:
I deleted this line, because I will not use the ProcessDialog class

elabcius5 committed:
Correction a service called

Acordding a Lucas's comment, I fixed a service calling and I deleted a
otto library and event bus. The service called was only a through
Manager class

juliocatano commented on app/src/main/java/com/globant/tutoria/subscriberssearch/api/SubscribersClient.java:
```diff
@@ -0,0 +1,46 @@
+public class SubscribersClient {
+
+    private static String API_URL = "https://api.github.com/repos/octokit/octokit.rb/";
+    private static volatile SubscribersClient subscribersClient;
```
Why do you need this volatile variable did you find any problem with thread safety?

juliocatano commented on app/src/main/java/com/globant/tutoria/subscriberssearch/api/SubscribersClient.java:
```diff
@@ -0,0 +1,46 @@
+    public static SubscribersClient getInstance() {
+        SubscribersClient localInstance = subscribersClient;
+        if (localInstance == null) {
+            synchronized (SubscribersClient.class) {
```
same as above, any thread safety issues? Or just implementing a thread safe singleton?

juliocatano commented on app/src/main/java/com/globant/tutoria/subscriberssearch/manager/SubscribersManager.java:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    public void getSubscribersList(){
+        sSubscribersClient.getSubscribersService().getSubscribers().enqueue(new Callback<List<SubscribersModel>>() {
```
You could also make the fragment implement the `Callback` interface

juliocatano commented on app/src/main/java/com/globant/tutoria/subscriberssearch/manager/SubscribersManager.java:
```diff
@@ -0,0 +1,43 @@
+        });
+    }
+
+    public interface SubscribersCallbackListener {
```
You could also make the fragment implement the `Callback` interface instead of creating one

juliocatano commented on app/src/main/res/layout/subscribers_list_fragment.xml:
```diff
@@ -0,0 +1,19 @@
+    android:layout_height="match_parent"
+    android:paddingBottom="@dimen/activity_vertical_margin"
+    android:paddingLeft="@dimen/activity_horizontal_margin"
+    android:paddingRight="@dimen/activity_horizontal_margin"
```
Please remove this paddings so the `RecyclerView` has all the space in the screen!
